### PR TITLE
Custom pass sg/2019.3

### PR DESCRIPTION
**Pull request by Paulm-Unity**
**DONT FORGET TO ADD A CHANGELOG**

### Checklist for PR maker
- Have you added a Label? : HDRP, Universal, ShaderGraph etc...
- Have you added a label for backport (if needed)? : need-backport-2019.3  .  When the PR is backported the label will be change ton backported-2019.3
- Have you added a changelog? Each package have a changelog.
- Have you updated or added the documentation for you PR? When property name is changed, when a feature behavior is change, when adding a new features, think to update the documentation in the same PR.
- Have you added a graphic test for your PR (if needed)? When adding new feature or discovering a bug that isn't cover by a test, please add a graphic test

---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?

---
### Testing status

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Comment by alelievr** [comment]
We might want to add also a node to get the _FadeValue and _CustomPassInjectionPoint values (as well as declare them into `ShaderVariablesCustomBlit.hlsl`) into the shader graph. Currently they are declared here: https://github.com/Unity-Technologies/ScriptableRenderPipeline/blob/master/com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPassCommon.hlsl#L10

File: com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl [hunk 1]
```diff
@@ -1,469 +1,477 @@
...
+// ----------------------------------------------------------------------------
+
+TEXTURE2D_X(_CameraDepthTexture);
+SAMPLER(sampler_CameraDepthTexture);
+
+// Color pyramid (width, height, lodcount, Unused)
+TEXTURE2D_X(_ColorPyramidTexture);
+
+// for custom blit shaders
+TEXTURE2D_X(_CustomBlitInput);
```
Comment (alelievr) [inline com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl]: Where in C# this texture is set ? Does users need to bind it manually in a custom C# fullscreen pass ?

**Comment by alelievr** [comment]
Another note, i don't remember what we said about the texture node, but i assume that we still want it with this PR right ?

Also i see that you have some line endings problem in your PR: all files you touched are changed entirely:
image
But if we hide whitespace changes it becomes this:
image

I guess there is some config problem, pushed files should always be in LF mode (git is supposed to convert automatically windows CRLF into LF when you push a file but it can be disabled)

File: com.unity.render-pipelines.high-definition/Editor/Material/CustomBlit/ShaderGraph/CreateCustomBlitShaderGraph.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+using UnityEditor.ShaderGraph;
+
+namespace UnityEditor.Rendering.HighDefinition
+{
+    static class CreateCustomBlitShaderGraph
+    {
+        [MenuItem("Assets/Create/Shader/HDRP/CustomBlit Graph", false, 208)]
+        public static void CreateMaterialGraph()
+        {
+            GraphUtil.CreateNewGraph(new CustomBlitMasterNode());
```
[the remark this answers is not in the document]
Comment (Paulm-Unity) [reply com.unity.render-pipelines.high-definition/Editor/Material/CustomBlit/ShaderGraph/CreateCustomBlitShaderGraph.cs]: I don't know. I would have to ask someone from the shader graph team.

File: com.unity.render-pipelines.high-definition/Editor/ShaderGraph/SharedCode.template.hlsl [hunk 3]
```diff
@@ -1,73 +1,77 @@
...
+        $SurfaceDescriptionInputs.WorldSpacePosition:        output.WorldSpacePosition =          input.positionRWS;
+        $SurfaceDescriptionInputs.ObjectSpacePosition:       output.ObjectSpacePosition =         TransformWorldToObject(input.positionRWS);
+        $SurfaceDescriptionInputs.ViewSpacePosition:         output.ViewSpacePosition =           TransformWorldToView(input.positionRWS);
+        $SurfaceDescriptionInputs.TangentSpacePosition:      output.TangentSpacePosition =        float3(0.0f, 0.0f, 0.0f);
+        $SurfaceDescriptionInputs.AbsoluteWorldSpacePosition:output.AbsoluteWorldSpacePosition =  GetAbsolutePositionWS(input.positionRWS);
+#if defined (BLIT_PASS)
+        $SurfaceDescriptionInputs.ScreenPosition:            output.ScreenPosition =              input.positionSS;
+#else
+        $SurfaceDescriptionInputs.ScreenPosition:            output.ScreenPosition =              ComputeScreenPos(TransformWorldToHClip(input.positionRWS), _ProjectionParams.x);
+#endif
```
[the remark this answers is not in the document]
Comment (Paulm-Unity) [reply com.unity.render-pipelines.high-definition/Editor/ShaderGraph/SharedCode.template.hlsl]: No, it was giving me incorrect results.

File: com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl (hunk 1 above)
Comment (Paulm-Unity) [reply com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl]: Yes it is set manually in C#

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs [hunk 4]
```diff
@@ -1,382 +1,393 @@
...
+        [SerializeField]
+        Version     m_Version = Version.Initial;
+        Version IVersionable<Version>.version
+        {
+            get => m_Version;
+            set => m_Version = value;
+        }
+
+
+        public RTHandle currentColorBuffer
```
[the remark this answers is not in the document]
Comment (Paulm-Unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs]: Ok I will change to use that function if it returns what I expect.

File: com.unity.render-pipelines.high-definition/Editor/ShaderGraph/SharedCode.template.hlsl (hunk 3 above)
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/ShaderGraph/SharedCode.template.hlsl]: Ok

File: com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl (hunk 1 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl]: Okay, maybe we should change a bit the design here to make it simpler to use. Here are my thought:
The `Texture Node` is supposed to provide all inputs we give in HDRP: `Normal`, `Roughness`, `Depth`, `Color`, `Custom Color` and `Custom Depth` (and we may add more in the future). This it is supposed to cover most cases of custom pass effects. If  the users want to create a more complex effect that will require them to allocate a new buffer themselves (which requires to create a C# custom pass script) and bind it to their ShaderGraph, They'll have to do irt through a blackboard parameter. This also allow them to have multiple custom targets at the same time.

What do you think about that ?

**Comment by Paulm-Unity** [comment]
Yes I am not sure how to solve this issue, I usually just hide whitespaces when looking at diffs.

File: com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl (hunk 1 above)
Comment (Paulm-Unity) [reply com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl]: Ok, so if I understand correctly you are talking about a new type of texture node that has a selector for some preset inputs? 

How would we bind a custom render target through a blackboard parameter though?

Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl]: Yes, IIRC it's the texture node we spoke about on this document: https://docs.google.com/document/d/1SOK6SsQhH6DvBby_K3f9O0M3sligYDygxBdceNXVejQ/edit

And for binding custom texture, I asked the SG team if we can expose a new blackboard type: `TEXTURE_X` that can either be tex2D or tex2DArray but they didn't answered me yet so we'll see. But for now users can already do it via a tex2DArray but there will be compatibility problems (on metal and xbox)

File: com.unity.render-pipelines.high-definition/Editor/Material/CustomBlit/ShaderGraph/CreateCustomBlitShaderGraph.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Paulm-Unity) [reply com.unity.render-pipelines.high-definition/Editor/Material/CustomBlit/ShaderGraph/CreateCustomBlitShaderGraph.cs]: Seems like this is not currently possible, but is on the SG team roadmap

**Pull request closed without merging**